Fix AR#== and AR#hash when new_record?

zenspider opened the pull request:
* Don't compute AR#== if new_record?
* Don't compute AR#hash if new_record?
* Use [array].hash for computing hash

### Motivation / Background

`AR#==` and `#hash` both check against `id` and then compute their values... but an instance that is a `new_record?` can have an `id` even if not saved if the schema has a default value associated with `id`.

cc @tenderlove to help backfill tests

### Detail

This fixes both methods to punt to `super` if `new_record?`.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [X] Tests are added or updated if you fix a bug or add a feature.
* [?] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

zenspider committed:
Fix AR#== and AR#hash when new_record?

* Don't compute AR#== if new_record?
* Don't compute AR#hash if new_record?
* Use [array].hash for computing hash

Co-authored-by: Aaron Patterson <tenderlove@ruby-lang.org>

tenderlove commented:
Looks like this breaks the build.  Judging from the tests that are broken, it seems like the tests aren't honoring the documentation.  [These tests assert that the hashes are the same](https://github.com/rails/rails/blob/946fc17c12c1509fd7e12c4a4c4a9848bff99c9a/activerecord/test/cases/core_test.rb#L182-L190) even though [the documentation on `eql?`](https://github.com/rails/rails/blob/946fc17c12c1509fd7e12c4a4c4a9848bff99c9a/activerecord/lib/active_record/core.rb#L543-L545) says that these should be different objects.

tenderlove committed:
Add a test for eql? on new objects

New objects shouldn't eql? other new objects even if the id column sets
the value to a default value

tenderlove committed:
Fix case when LHS is old record but RHS is new

nvasilevski commented:
> it seems like the tests aren't honoring the documentation.

Fair! These tests were added to mimic the behavior of a non-composite primary key model and a test like:
```ruby
  def test_two_new_records
    assert Topic.new(id: 1) == Topic.new(id: 1)
  end
```
would have failed as well, we just don't have one.

It also made sense if we treat Active Record objects as entities and two entities should be equal if they have the same identifier regardless of the persistence state. Though indeed it contradicts with the current docs.  

So I'm totally fine if we flip the assertion to align with our actual expectations on behavior for new records

zenspider commented:
@nvasilevski wanna add a commit here to realign the tests?

nvasilevski committed:
Flip eql? assertion for CPK new records

nvasilevski commented:
>wanna add a commit here to realign the tests?

Sure! I don't think I'm able to push the commit but here is the patch:

<details>
  <summary>Patch</summary>

  ```diff
    From 42e7e5df7fa3efa938db6cd49dfad0623cce75e3 Mon Sep 17 00:00:00 2001
    From: Nikita Vasilevsky <nikita.vasilevsky@shopify.com>
    Date: Wed, 5 Apr 2023 14:26:58 +0000
    Subject: [PATCH] Flip eql? assertion for CPK new records

    ---
    activerecord/test/cases/core_test.rb | 40 ++++++++++++++++++----------
    1 file changed, 26 insertions(+), 14 deletions(-)

    diff --git a/activerecord/test/cases/core_test.rb b/activerecord/test/cases/core_test.rb
    index f55b55ee5f5..e2750091977 100644
    --- a/activerecord/test/cases/core_test.rb
    +++ b/activerecord/test/cases/core_test.rb
    @@ -10,7 +10,7 @@ class NonExistentTable < ActiveRecord::Base; end
    class PkWithDefault < ActiveRecord::Base; end
    
    class CoreTest < ActiveRecord::TestCase
    -  fixtures :topics
    +  fixtures :topics, :cpk_books
    
    def test_eql_on_default_pk
        saved_record = PkWithDefault.new
    @@ -174,12 +174,12 @@ def test_find_by_cache_does_not_duplicate_entries
    
    def test_composite_pk_models_added_to_a_set
        library = Set.new
    -    # with primary key present
    +    # new record with primary key present
        library << Cpk::Book.new(author_id: 1, number: 2)
    
        # duplicate
    -    library << Cpk::Book.new(author_id: 1, number: 3)
    -    library << Cpk::Book.new(author_id: 1, number: 3)
    +    library << cpk_books(:cpk_great_author_first_book)
    +    library << cpk_books(:cpk_great_author_first_book)
    
        # without primary key being set
        library << Cpk::Book.new(title: "Book A")
    @@ -189,22 +189,34 @@ def test_composite_pk_models_added_to_a_set
    end
    
    def test_composite_pk_models_equality
    -    assert Cpk::Book.new(author_id: 1, number: 2) == Cpk::Book.new(author_id: 1, number: 2)
    +    book = cpk_books(:cpk_great_author_first_book)
    +    book_instance_1 = Cpk::Book.find_by(author_id: book.author_id, number: book.number)
    +    book_instance_2 = Cpk::Book.find_by(author_id: book.author_id, number: book.number)
    
    -    assert_not Cpk::Book.new(author_id: 1, number: 2) == Cpk::Book.new(author_id: 1, number: 3)
    +    assert book_instance_1 == book_instance_1
    +    assert book_instance_1 == book_instance_2
    +
    +    # two new records with the same primary key
    +    assert_not Cpk::Book.new(author_id: 1, number: 2) == Cpk::Book.new(author_id: 1, number: 2)
    +    # two new records with an empty primary key values
        assert_not Cpk::Book.new == Cpk::Book.new
    -    assert_not Cpk::Book.new(title: "Book A") == Cpk::Book.new(title: "Book B")
    -    assert_not Cpk::Book.new(author_id: 1) == Cpk::Book.new(author_id: 1)
    -    assert_not Cpk::Book.new(author_id: 1, title: "Same title") == Cpk::Book.new(author_id: 1, title: "Same title")
    +    # two persisted records with a different primary key
    +    assert_not cpk_books(:cpk_great_author_first_book) == cpk_books(:cpk_great_author_second_book)
    end
    
    def test_composite_pk_models_hash
    -    assert_equal Cpk::Book.new(author_id: 1, number: 2).hash, Cpk::Book.new(author_id: 1, number: 2).hash
    +    book = cpk_books(:cpk_great_author_first_book)
    +    book_instance_1 = Cpk::Book.find_by(author_id: book.author_id, number: book.number)
    +    book_instance_2 = Cpk::Book.find_by(author_id: book.author_id, number: book.number)
    
    -    assert_not_equal Cpk::Book.new(author_id: 1, number: 2).hash, Cpk::Book.new(author_id: 1, number: 3).hash
    +    assert_equal book_instance_1.hash, book_instance_1.hash
    +    assert_equal book_instance_1.hash, book_instance_2.hash
    +
    +    # two new records with the same primary key
    +    assert_not_equal Cpk::Book.new(author_id: 1, number: 2).hash, Cpk::Book.new(author_id: 1, number: 2).hash
    +    # two new records with an empty primary key values
        assert_not_equal Cpk::Book.new.hash, Cpk::Book.new.hash
    -    assert_not_equal Cpk::Book.new(title: "Book A").hash, Cpk::Book.new(title: "Book B").hash
    -    assert_not_equal Cpk::Book.new(author_id: 1).hash, Cpk::Book.new(author_id: 1).hash
    -    assert_not_equal Cpk::Book.new(author_id: 1, title: "Same title").hash, Cpk::Book.new(author_id: 1, title: "Same title").hash
    +    # two persisted records with a different primary key
    +    assert_not_equal cpk_books(:cpk_great_author_first_book).hash, cpk_books(:cpk_great_author_second_book).hash
    end
    end
    -- 
    2.32.0
  ```
</details>







[0001-Flip-eql-assertion-for-CPK-new-records.patch](https://github.com/rails/rails/files/11159910/0001-Flip-eql-assertion-for-CPK-new-records.patch)

tenderlove commented:
@nvasilevski thanks for writing the patch!  I've pulled it in to this branch

zenspider commented:
I've poked around at some of the failures left and tbh, they don't look valid to me. Either they need to create new records and _not_ expect them to be equal OR save them... but round tripping a new model through YAML? Not so much:

```
AttributeMethodsTest#test_YAML_dumping_a_record_with_time_zone-aware_attribute [/rails/activerecord/test/cases/attribute_methods_test.rb:817]:
No visible difference in the Topic#inspect output.
You should look at the implementation of #== on Topic or its members.
```

Open to suggestion

zenspider committed:
Try fixing model equality tests by saving a record first.

These tests were invalid per the equality doco for new records. Saving
one of the records _should_ allow it to travel the old path.

zenspider commented:
I've added a commit w/ 2 fixes and 1 skip to get to green.

I've got one failure due to saucelabs being dumb (surprise!)... not sure how to rerun failures on BK. But I think this should be considered green for our diffs.

zenspider commented:
I pushed another change and I still have a failure. These tests are hella flaky.

matthewd commented:
The documentation has been a lie since 4718d097ffe3af965f3ea7218156050507eabe4f, 15 minutes before it was added in a820d0afdd4a09e0902d7b7c4d9724f50089d254, 12 years ago. Switching to match the documentation is now a breaking change. :confused:

At this point I think this is quite well-established behaviour, and we should probably update the documentation to reflect reality instead, at least for `#==` (which has conceptually less strict equality semantics). 

While it's a change on `eql?`/`hash` too, it seems harder to argue surprise there: `==` is about values being functionally equivalent, while `eql?` is about them being functionally identical; treating two same-`id` new objects as "equivalent" is weird but plausible (we do so in some test assertions at GitHub), but claiming they're `eql?`-equal is more of a stretch.

---

Does switching to `[].hash` have a performance impact here? xor-folding lots of high-entropy values has clear problems, but I'd expect XORing a single high-entropy value with a single low-entropy value to maintain a reasonable distribution, and would have assumed it was better to avoid the array literal in such a hot path. (Or am I forgetting an opt that means it never gets allocated for that call?)

nvasilevski commented:
>(Or am I forgetting an opt that means it never gets allocated for that call?)

I believe this was the motivation:
https://github.com/ruby/ruby/blob/d83e59e6b8b52002cc46a14d7d4dc69416379029/hash.c#L311-L322

```
 * When implementing your own #hash based on multiple values, the best
 * practice is to combine the class and any values using the hash code of an
 * array:
 *
 * For example:
 *
 *   def hash
 *     [self.class, a, b, c].hash
 *   end
 *
 * The reason for this is that the Array#hash method already has logic for
 * safely and efficiently combining multiple hash values.
 * ```

eileencodes commented:
We hit this in our upgrade too and have a bunch of places that need to be fixed now. I think if we decide we do want this change it needs to be opt-in otherwise it's going to be a big headache once 7.1 is released. I think we revert this and discuss whether we want to make this change.

matthewd commented:
I suspect we can reasonably distinguish the default-value case from the explicitly-set-`id` case -- in the latter, it's reasonable to read the intention as being equal-ish objects. Having a static default on a primary key also seems unusual, but I can certainly buy an argument that `MyModel.new == MyModel.new` should always be false.

jhawthorn commented:
I also think we should revert this. It's going to cause significant upgrade friction for users (I know it has for us). If it's a change we want to make it needs at least warning and a deprecation cycle.

I would also argue that the documentation wasn't a lie, and did describe the existing behaviour, it's just unfortunately imprecise (and to be fair could be read either way). I read it as (my notes in bold):

> Returns true if [...] `comparison_object` is of the same type and `self` has an ID and it is equal to `comparison_object.id`.
>
> Note that new records are different from any other record by definition **[ie. using the definition above, assuming new records have no ID]**, unless the other record is the receiver itself. Besides, if you fetch existing records with `select` and leave the ID out, you're on your own, this predicate will return false **[note that: this is still just describing the effects of not having an ID]**

@zenspider can you elaborate on why your schema has a default for the primary key. Was there a motivation behind that or just something you ended up with?

tenderlove commented:
> Does switching to `[].hash` have a performance impact here?

Not right now, but it will when I [clean this up](https://github.com/ruby/ruby/pull/6090) it will (so it should be in Ruby 3.3).

> I suspect we can reasonably distinguish the default-value case from the explicitly-set-`id` case -- in the latter, it's reasonable to read the intention as being equal-ish objects.

If we can do this, that would be great.  I'm fine if we revert this but I'd really like to keep some of the explicit `.hash` and `.eql?` tests because we clearly weren't explicitly testing their behavior.

Hopefully @zenspider can give more context, but from what I understand they're making new objects from an association, the new objects get the "default pk" value, Rails puts the objects in to a `Set` somehow, and then all but one of the new instances disappear because they all hash the same.  An example using associations would be instructive I think.

tenderlove commented:
> I'm fine if we revert this but I'd really like to keep some of the explicit `.hash` and `.eql?` tests because we clearly weren't explicitly testing their behavior.

Actually all of the tests I added use the table that sets a default value on the `id` column.  Maybe we should just revert the whole thing? 😫

zenspider commented:
Yes, I believe the problem stems from this code:

https://github.com/rails/rails/blob/main/activerecord/lib/active_record/associations/collection_association.rb#L466-L473

with newly built relations on a table with a default `0` on `id` (don't ask... I have no idea, but apparently this is kosher on mysql)

so reverting this code will bring this problem back to life. I think there needs to be an extra test ensuring that a new model instance can build more than one relation instance if id=0 for all relation instances. 

I will _also_ attempt to get our DBAs to rectify the schema, but something something snowballs.

jhawthorn commented:
Thanks Ryan. I'm still a bit confused. If we were to patch around that behaviour in collection association, aren't those records then unsaveable (because they all have the same primary key id).

I noticed you can't have a default and auto_increment:
```
mysql> create table hawthtest ( `id` int DEFAULT(0) AUTO_INCREMENT );
ERROR 1067 (42000): Invalid default value for 'id'
```

However if you don't auto-increment, it does seem to be allowed (whether or not it is a primary key):
```
mysql> create table hawthtest ( `id` int DEFAULT(0) );
Query OK, 0 rows affected (0.22 sec)

mysql> create table hawthtest2 ( `id` int DEFAULT(0), PRIMARY KEY(`id`) );
Query OK, 0 rows affected (0.20 sec)
```

Without auto_increment, I can't imagine we're able to save those records (unless an id is explicitly specified?). And if the id is non-unique we'll certainly have problems everywhere (essentially the same as this issue but with records _from_ the DB). So I'm having a hard time seeing what the goal and working state is after were to work around the initial problem in the association. But maybe there's something I've missed in your schema/setup.

zenspider commented:
dood... I'm not a dba. So I'm guessing at best. (I'm asking the DBAs but ... timezones)

```
CREATE TABLE `taggings` (
  `id` bigint(20) NOT NULL DEFAULT '0',
...
  PRIMARY KEY (`id`),
...
) ENGINE=InnoDB;
```

AH. it looks like this bit: 

> No value was specified for the AUTO_INCREMENT column, so MySQL assigned sequence numbers automatically. You can also explicitly assign 0 to the column to generate sequence numbers, ...

https://dev.mysql.com/doc/refman/8.0/en/example-auto-increment.html

tenderlove commented:
We can't really ignore the default value of `0` because that would break saving the model.  Maybe we could treat `0` as "not set" for mysql?

matthewd commented:
> Maybe we could treat 0 as "not set" for mysql?

Yeah, we know it's an autoincrement column, and that zero behaviour is well-documented... I think we're currently (and fixably) wrong to believe/report that the value will be zero.
